### PR TITLE
Custom page poc

### PR DESCRIPTION
**Pull request by harterjwustl**

### CODE REVIEW

File: coldfront/plugins/qumulo/api/usage/usage_allocations.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            )
+
+            allocations = (
+                pi_allocations | billing_allocations | technical_allocations
+            ).distinct()
+
+        allocation_path_map = map(
+            lambda allocation: {
+                "id": allocation[0],
+                "path": allocation[1],
```
Comment (melonistic) [inline coldfront/plugins/qumulo/api/usage/usage_allocations.py]: @harterjwustl are there any reasons for not using `storage_filesystem_path` instead of the generic `path`?

**Comment by cdeanj2** [comment]
Overall this looks good, but hesitant to approve because of failing tests.

File: coldfront/config/plugins/reporting.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+INSTALLED_APPS += [
+    "coldfront.plugins.reporting",
+]
+
+STATICFILES_DIRS += [
+    PROJECT_ROOT("coldfront/plugins/reporting/frontend/react/dist"),
+]
+
+TEMPLATES[0]["DIRS"] += [PROJECT_ROOT("coldfront/plugins/reporting/templates")]
```
Comment (melonistic) [inline coldfront/config/plugins/reporting.py]: @harterjwustl this configuration should be unnecessary if the templates are placed according to the Django convention. BTW, the [iQuota plugin templates](https://github.com/WashU-IT-RIS/coldfront-wustl-fork/tree/main/coldfront/plugins/iquota/templates/iquota) follow the Django convention.

File: coldfront/config/plugins/reporting.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    PROJECT_ROOT,
+    STATICFILES_DIRS,
+)
+
+INSTALLED_APPS += [
+    "coldfront.plugins.reporting",
+]
+
+STATICFILES_DIRS += [
+    PROJECT_ROOT("coldfront/plugins/reporting/frontend/react/dist"),
```
Comment (melonistic) [inline coldfront/config/plugins/reporting.py]: @harterjwustl should this be "coldfront/plugins/reporting/static/..." ?

Comment (harterjwustl) [reply coldfront/config/plugins/reporting.py]: No, this is where built frontend files are put, and django needs to be alerted to include this location

File: coldfront/plugins/qumulo/api/usage/usage_allocations.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            AllocationAttribute.objects.filter(
+                allocation=OuterRef("pk"),
+                allocation_attribute_type__name="storage_filesystem_path",
+            ).values("value")
+        )
+
+        allocations = (
+            Allocation.objects.parents()
+            .annotate(storage_filesystem_path=storage_filesystem_path_query)
+            .filter(resources__resource_type__name="Storage", status__name="Active")
```
Comment (melonistic) [inline coldfront/plugins/qumulo/api/usage/usage_allocations.py]: @harterjwustl use the `active_storage` query set instead?

```
Allocation.objects.parent().active_storage()
```

File: coldfront/plugins/qumulo/api/usage/usages.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+
+class Usages(LoginRequiredMixin, UserPassesTestMixin, View):
+    def test_func(self):
+        user: User = self.request.user
+        if user.is_superuser or user.is_staff:
+            return True
+
+        allocation_pk = int(self.request.GET.get("allocation_id"))
+        allocation = Allocation.objects.get(pk=allocation_pk)
```
Comment (melonistic) [inline coldfront/plugins/qumulo/api/usage/usages.py]: @harterjwustl this is a good candidate to move to a permissions service.

File: coldfront/plugins/qumulo/api/usage/usages.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        end_date_usage: (
+            AllocationAttributeUsage
+        ) = AllocationAttributeUsage.history.as_of(end_datetime).get(
+            allocation_attribute__allocation=allocation,
+            allocation_attribute__allocation_attribute_type__name="storage_quota",
+        )
+        end_date_quota: AllocationAttribute = (
+            end_date_usage.allocation_attribute.history.as_of(end_datetime)
+        )
+        usage_gib.append(
```
Comment (melonistic) [inline coldfront/plugins/qumulo/api/usage/usages.py]: @harterjwustl what is gib?

Comment (harterjwustl) [reply coldfront/plugins/qumulo/api/usage/usages.py]: gibibyte

Comment (melonistic) [reply coldfront/plugins/qumulo/api/usage/usages.py]: gibibytes!

File: coldfront/plugins/qumulo/tests/api/usage/usage_allocations/test_usage_allocations_get.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+            allocations["storage_allocation"].status = AllocationStatusChoiceFactory(
+                name="Pending"
+            )
+            allocations["storage_allocation"].save()
+
+        self.client.force_login(user)
+        response = self.client.get("/qumulo/api/usage/allocations")
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.json()["allocations"]), expected_allocation_count)
```
Comment (melonistic) [inline coldfront/plugins/qumulo/tests/api/usage/usage_allocations/test_usage_allocations_get.py]: @harterjwustl Would you consider including a test showing that the result set does not have any of the pending allocations?

File: coldfront/plugins/qumulo/tests/api/usage/usage_allocations/test_usage_allocations_get.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+            billing_contact_attribute = AllocationAttribute.objects.get(
+                allocation__pk=allocations["storage_allocation"].pk,
+                allocation_attribute_type__name="billing_contact",
+            )
+            billing_contact_attribute.value = user.username
+            billing_contact_attribute.save()
+
+        for _ in range(expected_allocation_count):
+            end_path = fake.last_name()
+            create_ris_project_and_allocations_storage3(f"/storage3/fs1/{end_path}")
```
Comment (melonistic) [inline coldfront/plugins/qumulo/tests/api/usage/usage_allocations/test_usage_allocations_get.py]: @harterjwustl just like the test above, would you consider adding test showing that none of these allocations are in the result set?

File: coldfront/plugins/qumulo/tests/api/usage/usage_allocations/test_usage_allocations_get.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+            create_allocation_linkage(
+                parent_allocations["storage_allocation"],
+                [child_allocations["storage_allocation"]],
+            )
+
+        self.client.force_login(user)
+        response = self.client.get("/qumulo/api/usage/allocations")
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.json()["allocations"]), expected_allocation_count)
```
Comment (melonistic) [inline coldfront/plugins/qumulo/tests/api/usage/usage_allocations/test_usage_allocations_get.py]: @harterjwustl would you show that the child allocation is not in the result set?

**Review by melonistic — COMMENTED** [review]
I reviewed the Python code. Meanwhile, I am not well versed with React, so I did a cursory review. I would like someone with better React skills to review.

File: coldfront/plugins/qumulo/api/usage/usage_allocations.py (hunk 1 above)
Comment (harterjwustl) [reply coldfront/plugins/qumulo/api/usage/usage_allocations.py]: I decided that I preferred that the api delivered content that was somewhat decoupled/agnostic of the plugin's structure, as it is being used outside of the plugin and the normal coldfront workflow, and simply using `path` seemed sufficient

File: coldfront/plugins/qumulo/api/usage/usages.py (hunk 5 above)
Comment (harterjwustl) [reply coldfront/plugins/qumulo/api/usage/usages.py]: I'm going to need some more details on this, but if you mean moving it into a services class...I'm going to say it's a YAGNI issue until we want to use it like this again.  Right now it's a one off permissions function that's using a feature that already is extant in Django for custom access requirements.

I'd rather not refactor the tests and code here to genericize into a service class if it isn't needed.

File: coldfront/plugins/qumulo/tests/api/usage/usage_allocations/test_usage_allocations_get.py
Comment (harterjwustl) [inline coldfront/plugins/qumulo/tests/api/usage/usage_allocations/test_usage_allocations_get.py]: @melonistic , I've gone ahead a refactored a number of the tests here as requested, so that we are cross checking the response lists against expected ids to definitively prove that no unwanted results are returned

File: coldfront/plugins/qumulo/tests/api/usage/usage_allocations/test_usage_allocations_get.py [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+
+        for id in expected_allocation_ids:
+            response_ids = map(
+                lambda response_allocation: response_allocation["id"],
+                response_allocations,
+            )
+            self.assertIn(id, response_ids)
+
+    def test_multiple_cases_get_their_allocations(self):
+        allocatoion_iterator = 5
```
Comment (melonistic) [inline coldfront/plugins/qumulo/tests/api/usage/usage_allocations/test_usage_allocations_get.py]: @harterjwustl [Trivial] Spelling and this is not an iterator. I would sayallocation_count or something like that.

**Review by melonistic — APPROVED** [review]
Ship it!
